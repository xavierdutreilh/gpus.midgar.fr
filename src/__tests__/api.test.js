const request = require("supertest")

const api = require("../api")

describe("GET /", () => {
  describe("Accept: text/html", () => {
    it("redirects to /offers", async () => {
      const response = await request(api.callback())
        .get("/")
        .set("Accept", "text/html")
      expect(response.status).toBe(302)
      expect(response.headers.location).toEqual("/offers")
      expect(response.text).toMatchSnapshot()
    })
  })

  describe("Accept: application/json", () => {
    it("retrieves system information", async () => {
      const response = await request(api.callback())
        .get("/")
        .set("Accept", "application/json")
      expect(response.status).toBe(200)
      expect(response.body).toEqual({
        name: "gpus.midgar.fr",
        version: expect.any(String),
        environment: "test",
      })
    })
  })
})

describe("GET /offers", () => {
  describe("Accept: text/html", () => {
    it("renders offers", async () => {
      const response = await request(api.callback())
        .get("/offers")
        .set("Accept", "text/html")
      expect(response.status).toBe(200)
      expect(response.text).toMatchSnapshot()
    })
  })

  describe("Accept: application/json", () => {
    it("retrieves offers", async () => {
      const response = await request(api.callback())
        .get("/offers")
        .set("Accept", "application/json")
      expect(response.status).toBe(200)
      expect(response.body).toEqual([
        {
          store: "nvidia",
          name: "NVIDIA GEFORCE RTX 3080",
          price: "€719.00",
          status: "unavailable",
          url: null,
        },
        {
          store: "nvidia",
          name: "NVIDIA GEFORCE RTX 3090",
          price: "€1,549.00",
          status: "unavailable",
          url: null,
        },
        {
          store: "ldlc",
          name: "ASUS GeForce ROG STRIX RTX 3080 O10G GAMING",
          price: "€1,099.95",
          status: "unavailable",
          url: "https://www.ldlc.com/fiche/PB00375431.html",
        },
        {
          store: "ldlc",
          name: "ASUS GeForce ROG STRIX RTX 3090 O24G GAMING",
          price: "€1,949.95",
          status: "unavailable",
          url: "https://www.ldlc.com/fiche/PB00369849.html",
        },
      ])
    })
  })
})

describe("GET /offers?store&name", () => {
  describe("Accept: text/html", () => {
    it("renders offers", async () => {
      const response = await request(api.callback())
        .get("/offers?store=nvidia&name=rtx+3090")
        .set("Accept", "text/html")
      expect(response.status).toBe(200)
      expect(response.text).toMatchSnapshot()
    })
  })

  describe("Accept: application/json", () => {
    it("retrieves offers", async () => {
      const response = await request(api.callback())
        .get("/offers?store=nvidia&name=rtx+3090")
        .set("Accept", "application/json")
      expect(response.status).toBe(200)
      expect(response.body).toEqual([
        {
          store: "nvidia",
          name: "NVIDIA GEFORCE RTX 3090",
          price: "€1,549.00",
          status: "unavailable",
          url: null,
        },
      ])
    })
  })
})

describe("GET /assets/application.css", () => {
  it("retrieves styles", async () => {
    const response = await request(api.callback()).get(
      "/assets/application.css"
    )
    expect(response.status).toBe(200)
    expect(response.text).toMatchSnapshot()
  })
})
