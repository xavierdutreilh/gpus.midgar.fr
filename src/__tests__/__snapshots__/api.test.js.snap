// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET / Accept: text/html redirects to /offers 1`] = `"Redirecting to <a href=\\"/offers\\">/offers</a>."`;

exports[`GET /offers Accept: text/html renders offers 1`] = `"<!DOCTYPE html><html lang=\\"en\\"><head><meta charset=\\"utf-8\\"><meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\"><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, shrink-to-fit=no\\"><title>Offers | gpus.midgar.fr</title></head><body><main><form><label for=\\"store\\">Store</label><select id=\\"store\\" name=\\"store\\"><option value=\\"\\">All</option><option value=\\"ldlc\\">LDLC</option><option value=\\"materielnet\\">Materiel.net</option><option value=\\"nvidia\\">Nvidia</option></select><label for=\\"name\\">Name</label><input type=\\"text\\" id=\\"name\\" name=\\"name\\"><label for=\\"status\\">Status</label><select id=\\"status\\" name=\\"status\\"><option value=\\"\\">All</option><option value=\\"available\\">Available</option><option value=\\"unavailable\\">Unavailable</option><option value=\\"unknown\\">Unknown</option></select><button type=\\"submit\\">Search</button></form><table><thead><tr><th>Store</th><th>Name</th><th>Price</th><th>Status</th></tr></thead><tbody><tr><td>nvidia</td><td>NVIDIA GEFORCE RTX 3080</td><td>€719.00</td><td>unavailable</td></tr><tr><td>nvidia</td><td>NVIDIA GEFORCE RTX 3090</td><td>€1,549.00</td><td>unavailable</td></tr><tr><td>ldlc</td><td><a href=\\"https://www.ldlc.com/fiche/PB00375431.html\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">ASUS GeForce ROG STRIX RTX 3080 O10G GAMING</a></td><td>€1,099.95</td><td>unavailable</td></tr><tr><td>ldlc</td><td><a href=\\"https://www.ldlc.com/fiche/PB00369849.html\\" target=\\"_blank\\" rel=\\"noopener noreferrer\\">ASUS GeForce ROG STRIX RTX 3090 O24G GAMING</a></td><td>€1,949.95</td><td>unavailable</td></tr></tbody></table></main></body></html>"`;

exports[`GET /offers?store&name Accept: text/html renders offers 1`] = `"<!DOCTYPE html><html lang=\\"en\\"><head><meta charset=\\"utf-8\\"><meta http-equiv=\\"x-ua-compatible\\" content=\\"ie=edge\\"><meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1, shrink-to-fit=no\\"><title>Offers | gpus.midgar.fr</title></head><body><main><form><label for=\\"store\\">Store</label><select id=\\"store\\" name=\\"store\\"><option value=\\"\\">All</option><option value=\\"ldlc\\">LDLC</option><option value=\\"materielnet\\">Materiel.net</option><option value=\\"nvidia\\" selected>Nvidia</option></select><label for=\\"name\\">Name</label><input type=\\"text\\" id=\\"name\\" name=\\"name\\" value=\\"rtx 3090\\"><label for=\\"status\\">Status</label><select id=\\"status\\" name=\\"status\\"><option value=\\"\\">All</option><option value=\\"available\\">Available</option><option value=\\"unavailable\\">Unavailable</option><option value=\\"unknown\\">Unknown</option></select><button type=\\"submit\\">Search</button></form><table><thead><tr><th>Store</th><th>Name</th><th>Price</th><th>Status</th></tr></thead><tbody><tr><td>nvidia</td><td>NVIDIA GEFORCE RTX 3090</td><td>€1,549.00</td><td>unavailable</td></tr></tbody></table></main></body></html>"`;
